src/controller/java/CHIPDeviceController-JNI.cpp

```diff
@@ -1357,6 +1358,130 @@ JNI_METHOD(void, read)
+    SuccessOrExit(JniReferences::GetInstance().GetListSize(attributeList, listSize));
+
+    writeClient = Platform::New<app::WriteClient>(device->GetExchangeManager(), callback->GetChunkedWriteCallback(),
+                                                  timedRequestTimeoutMs != 0 ? Optional<uint16_t>(timedRequestTimeoutMs)
+                                                                             : Optional<uint16_t>::Missing());
+
+    for (uint8_t i = 0; i < listSize; i++)
+    {
+        jobject attributeItem             = nullptr;
+        uint32_t endpointId               = 0;
+        uint32_t clusterId                = 0;
+        uint32_t attributeId              = 0;
+        jmethodID getEndpointIdMethod     = nullptr;
+        jmethodID getClusterIdMethod      = nullptr;
+        jmethodID getAttributeIdMethod    = nullptr;
+        jmethodID hasDataVersionMethod    = nullptr;
+        jmethodID getDataVersionMethod    = nullptr;
+        jmethodID getTlvByteArrayMethod   = nullptr;
+        jobject endpointIdObj             = nullptr;
+        jobject clusterIdObj              = nullptr;
+        jobject attributeIdObj            = nullptr;
+        jbyteArray tlvBytesObj            = nullptr;
+        bool hasDataVersion               = false;
+        Optional<DataVersion> dataVersion = Optional.Empty();
+        jbyte * tlvBytesObjBytes          = nullptr;
+        jsize length                      = 0;
+        TLV::TLVReader reader;
+
+        SuccessOrExit(JniReferences::GetInstance().GetListItem(attributeList, i, attributeItem));
+        SuccessOrExit(JniReferences::GetInstance().FindMethod(env, attributeItem, "getEndpointId",
+                                                              "()Lchip/devicecontroller/model/ChipPathId;", &getEndpointIdMethod));
+        SuccessOrExit(JniReferences::GetInstance().FindMethod(env, attributeItem, "getClusterId",
+                                                              "()Lchip/devicecontroller/model/ChipPathId;", &getClusterIdMethod));
+        SuccessOrExit(JniReferences::GetInstance().FindMethod(env, attributeItem, "getAttributeId",
+                                                              "()Lchip/devicecontroller/model/ChipPathId;", &getAttributeIdMethod));
+        SuccessOrExit(JniReferences::GetInstance().FindMethod(env, attributeItem, "hasDataVersion", "()Z", &hasDataVersionMethod));
+        SuccessOrExit(JniReferences::GetInstance().FindMethod(env, attributeItem, "getDataVersion", "()I", &getDataVersionMethod));
+        SuccessOrExit(
+            JniReferences::GetInstance().FindMethod(env, attributeItem, "getTlvByteArray", "()[B", &getTlvByteArrayMethod));
+
+        endpointIdObj = env->CallObjectMethod(attributeItem, getEndpointIdMethod);
+        VerifyOrExit(!env->ExceptionCheck(), err = CHIP_JNI_ERROR_EXCEPTION_THROWN);
+        VerifyOrExit(endpointIdObj != nullptr, err = CHIP_ERROR_INCORRECT_STATE);
```

Comment:
Should these be `CHIP_ERROR_INVALID_ARGUMENT`? Seems like this is validating the incoming request?

Reply:
updated, thanks